[FEAT] 정산 관련 API 구현

Chominsung0823 opened the pull request:
## 🪺 개요 
<img width="605" height="221" alt="image" src="https://github.com/user-attachments/assets/f9442b4f-736b-47ee-bbab-616dbb1da59d" />

정산 관련 API 4가지를 개발하였습니다.

## 💻 작업 사항 
1. 정산 목록 조회 API 구현
- 사건 중 완료된 사건만 조회하도록 조건을 적용하였습니다.
- 사건(case) 테이블과 정산(settlement) 테이블을 LEFT JOIN하여, 정산 정보가 없는 사건도 목록에 포함되도록 설계하였습니다.
- JOIN 이후 정산 데이터가 존재하지 않을 경우, 프론트 구분을 위해 정산 정보 필드는 null로 반환되도록 처리하였습니다.

2. 정산 상세 조회 API 구현
- 정산 ID 기준으로 단일 정산 정보를 반환하도록 설계했습니다.
- 정산 정보가 존재하지 않을 경우를 위해 예외 처리를 추가하였습니다.

3. 정산 등록 API 구현
- 신규 정산 데이터를 생성하여 저장합니다.
- VAT 포함 여부, 변호사 보수, 총 청구액 등 금액 필드를 검증합니다.

4. 정산 수정 API 구현
- 정산서를 수정할 수 있는 API를 구현했습니다.

##  변경사항 추가(11/19)
- CreateSettlementRequest에서 paymentStatus 제거 및 CaseErrorCode 사용
- 정산서 결제 상태 변경 API 구현, UpdatePaymentStatusRequest DTO 추가
- 정산서 상태 관리(settlement_status) 추가, NONE(미작성)→DRAFT(작성중)→FINAL(작성완료) 자동 전환 로직 구현
- 정산 필드명 개선 (attorneyFee, agreementAmount 등)
- 정산 목록 조회 범위를 "진행중" 사건까지 확장
- 의뢰인 수령액(clientReceivable) 요청 제거 후 내부 자동 계산

> replit 기준으로 구현하였고, 수정 부분도 필요할 거 같아서 추가하였습니다.

## 🌱 Issue Number
- #9 

## ✅ 셀프 체크리스트
- [x] PR 제목을 형식에 맞게 작성했나요 ?
- [x] 브랜치 전략에 맞는 브랜치에 PR을 올리고 있나요 ?
- [x] 테스트는 잘 통과했나요 ?
- [x] 빌드에 성공했나요 ?
- [x] 본인을 Assign 해주세요.
- [x] 리뷰할 팀원들을 Request 해주세요.
- [x] 해당 PR에 맞는 label을 붙여주세요.

## 🙏 To Reviewers
>  API 기능 별로 Commit 하지 않았고, 연관된 File 별로 Commit 했습니다. 다음부턴 기능 별로 구현할 때마다 Commit 하겠습니다!!

Chominsung0823 committed:
feat: 정산 도메인 엔티티 및 Enum 작성 (#9)

- Settlement 엔티티 생성
- PaymentStatus Enum 추가
- BaseEntity 상속 및 공통 필드 적용

Chominsung0823 committed:
feat: 정산 관련 Request/Response DTO 작성 (#9)

- CreateSettlementRequest, UpdateSettlementRequest 작성
- SettlementResponse, SettlementListResponse 작성

Chominsung0823 committed:
feat: 정산 Repository 작성 (#9)

- SettlementRepository 생성
- 완료된 사건 기준 정산 리스트 조회 쿼리 작성
- Object[] 결과를 반환하도록 설정

Chominsung0823 committed:
feat: 정산 Service 구현 및 예외 처리 추가 (#9)

- 정산 생성(createSettlement) 구현
- 정산 목록 조회(getSettlements) 구현
- 정산서 상세 조회(getSettlement) 구현
- 정산 수정(updateSettlement) 구현
- 트랜잭션 처리 및 SettlementException 적용

Chominsung0823 committed:
feat: 정산 Controller 작성 (#9)

- SettlementController 생성
- 정산 생성, 목록 조회, 상세 조회, 수정 API 구현

Chominsung0823 committed:
Merge branch 'develop' into feat/settlement-#9

Chominsung0823 committed:
refactor: 정산 방식 수정 (#9)

- 의뢰인 수령액 = 합의금 총액 - (변호사비 + 실비 + VAT(선택)) 형태로 수정

Chominsung0823 committed:
refactor: 정산 엔티티의 사건 연관 필드를 cases로 변경하고 1대1 관계로 재설계 (#9)

erika0915 commented on src/main/java/com/haeil/be/settlement/dto/request/CreateSettlementRequest.java:
```diff
@@ -0,0 +1,65 @@
+@NoArgsConstructor
+public class CreateSettlementRequest {
+
+    @NotNull(message = "결제 상태는 필수 입력 항목입니다.")
```
결제 상태를 request로 주는 이유가 무엇인가요 ? 결제 상태는 서비스 레이어에서 처리할 수 있는 게 아닌가용 ? 혹시 입금 대기에서 입금 완료 상태로 바꾸는 거라면 그거는 뭔가 api가 하나가 필요할 것 같아요 ! (근데 생각해보니깐 수임에서도 하나 만들었어야 했는데 !! ㅎㅎ ㅠㅠ 저도 api 하나 만들게용)

Chominsung0823 replied:
정산을 생성할 때 정산 상태가 "입금대기"로 초기 설정 되니까 request에 있으면 안 되겠네요! 상태 변경 API 추가하겠습니다!!

erika0915 commented on src/main/java/com/haeil/be/settlement/service/SettlementService.java:
```diff
@@ -0,0 +1,134 @@
+                                () ->
+                                        new SettlementException(
+                                                SettlementErrorCode.SETTLEMENT_NOT_FOUND));
+
```
사소한거긴한데 .. Case를 찾는건데 SETTLEMENT_NOT_FOUND 에러 보다 CaseErrorCode에 CASE_NOT_FOUND가 더 적합할 것 같아요 !

Chominsung0823 replied:
넵 수정하겠습니다!!

erika0915 commented on src/main/java/com/haeil/be/settlement/controller/SettlementController.java:
```diff
@@ -0,0 +1,71 @@
+        return ApiResponse.from(response);
+    }
+
+    @Operation(summary = "정산 수정", description = "정산 정보를 수정합니다.")
```
정산 수정은 어느 경우에 필요한 걸까요 ? 연결하는 부분이 있나용 ?

Chominsung0823 replied:
정산서를 작성한 이후에도 실비나 변호사 보수 등 금액이 추가로 발생할 수 있다고 생각했습니다.
그래서 정산서 상세 화면에서 필요한 경우 수정할 수 있도록 기능을 추가했습니다!

erika0915 commented on src/main/java/com/haeil/be/settlement/service/SettlementService.java:
```diff
@@ -0,0 +1,134 @@
+    public List<SettlementListResponse> getSettlements() {
+        List<Object[]> results =
+                settlementRepository.findSettlementsByCaseStatus(CaseStatus.COMPLETED);
+
```
정산 과정이 사건이 '종료' 된 것들에 대해서 정산을 하나요 ? 그 사건에서 종결 보고서를 작성해야 '종료'로 되는 거는 아닌가용! 그래서 그 진행중인 사건들에 대해서 정산서를 작성하는 거라고 생각했거든요 !  어떻게 되는건가용 ?

Chominsung0823 replied:
사건이 진행 중일 때도 비용 발생 시 정산서를 작성할 수 있겠네요.
그러면 변호사 비용이나 실비 같은 필드들을 null 값을 허용하고, 계산 할 때 null이라면 0으로 처리하도록 할게요!
또 진행중인 사건에 대해 정산서를 작성하는 거라면 수정 API가 필요해보입니다!

yjy323 replied:
results.stream().map(SettlementListResponse::from).collect(Collectors.toList()); 관련한 질문인데,

Repository에서 JOIN 쿼리를 지정하면 Object[]에 인스턴스가 각각 저장되나보네요?! 처음 알아서 배워가요~~

erika0915 commented on src/main/java/com/haeil/be/settlement/repository/SettlementRepository.java:
```diff
@@ -0,0 +1,24 @@
+     */
+    @Query(
+            "SELECT c, s FROM com.haeil.be.cases.domain.Cases c "
+                    + "LEFT JOIN com.haeil.be.settlement.domain.Settlement s ON s.cases.id = c.id "
```
jpql로 left join하는거 좋은 것 같아요~👍🏻

Chominsung0823 committed:
refactor: 정산 생성 로직 개선 (#9)

- CreateSettlementRequest DTO에서 paymentStatus 필드 제거
- 사건 미조회 시 CaseErrorCode.CASE_NOT_FOUND 사용

yjy323 reviewed (approved):
수희님이 리뷰 잘 남겨 주시고, 민성님도 바로 확인해주셔서 제가 추가로 적어야 할 부분은 없었습니다!
수고하셨습니다~~

Chominsung0823 committed:
feat: 결제 상태 변경 API 구현 (#9)

- 정산서의 결제 상태를 별도 API로 변경할 수 있도록 구현
- UpdatePaymentStatusRequest DTO 생성
- update와 연결된 결제 상태(paymentStatus) 필드 제거

yangjunmo131 reviewed (approved):
구현하느라 수고하셨습니다!!

Chominsung0823 committed:
feat: 정산서 상태 관리 추가 및 필드명 개선 (#9)

- 정산서 작성 상태 관리를 위해 settlement_status 필드 추가
- NONE → DRAFT(임시작성) → FINAL(작성완료) 자동 전환 로직 구현
- 생성/수정 시 settlement 엔티티 내부에서 상태 자동 업데이트
- 정산 관련 필드명 의미가 명확하도록 역할 기반으로 변경

Chominsung0823 committed:
refactor: 정산 목록 조회 범위 확장 및 의뢰인 수령액(clientReceivable) 자동 계산 (#9)

- 정산 목록 조회 시 완료된 사건뿐만 아니라 진행중 사건도 포함
- 의뢰인 수령액(clientReceivable)을 요청에서 제거하고, Settlement 내부에서 자동 계산하도록 수정

erika0915 reviewed (approved):
확인했습니당 ~ 고생많으셨어요 ! 👍🏻💯
